src/coreclr/vm/methodtable.cpp

```diff
@@ -2903,6 +2928,311 @@ int MethodTable::GetRiscV64PassStructInRegisterFlags(TypeHandle th)
+            NativeFieldCategory category = fields[i].GetCategory();
+            if (category == NativeFieldCategory::NESTED)
+            {
+                int elementTypeIndex = typeIndex;
+
+                MethodTable* nested = fields[i].GetNestedNativeMethodTable();
+                if (!FlattenFields(TypeHandle(nested), offset + fields[i].GetExternalOffset(), info, typeIndex DEBUG_ARG(nestingLevel + 1)))
+                    return false;
+
+                // In native layout fixed arrays are marked as NESTED just like structs
+                int nElements = fields[i].GetNumElements();
+                if (!HandleInlineArray(elementTypeIndex, nElements, info, typeIndex DEBUG_ARG(nestingLevel + 1)))
+                    return false;
+            }
+            else if (fields[i].NativeSize() <= TARGET_POINTER_SIZE)
+            {
+                if (typeIndex >= 2)
+                {
+                    LOG((LF_JIT, LL_EVERYTHING, "FpStructInRegistersInfo:%*s  * too many fields\n",
+                        nestingLevel * 4, ""));
+                    return false;
+                }
+
+                SetFpStructInRegistersInfoField(info, typeIndex++,
+                    (category == NativeFieldCategory::FLOAT),
+                    FpStruct::IntKind::Signed, // NativeFieldDescriptor doesn't save signedness, TODO: should it?
+                    fields[i].NativeSize(),
+                    offset + fields[i].GetExternalOffset());
+                LOG((LF_JIT, LL_EVERYTHING, "FpStructInRegistersInfo:%*s  * found field %s [%i..%i), type: %s\n",
+                    nestingLevel * 4, "", fields[i].GetFieldDesc()->GetDebugName(),
+                    fields[i].GetExternalOffset(), fields[i].GetExternalOffset() + fields[i].NativeSize(), categoryNames[(int)category]));
+            }
+            else
+            {
+                LOG((LF_JIT, LL_EVERYTHING, "FpStructInRegistersInfo:%*s "
+                    " * field %s, type: %s, is too big (%i bytes)\n",
+                    nestingLevel * 4, "", fields[i].GetFieldDesc()->GetDebugName(),
+                    categoryNames[(int)category], fields[i].NativeSize()));
+                return false;
+            }
+        }
+    }
+    return true;
+}
+#endif
+
+#if defined(TARGET_RISCV64)
+
+static FpStructInRegistersInfo GetRiscV64PassFpStructInRegistersInfoImpl(TypeHandle th)
+{
+    FpStructInRegistersInfo info = {};
+    int nFields = 0;
+    if (!FlattenFields(th, 0, info, nFields DEBUG_ARG(0)))
+        return FpStructInRegistersInfo{};
+
+    using namespace FpStruct;
+    if ((info.flags & (FloatInt | IntFloat)) == 0)
+    {
+        LOG((LF_JIT, LL_EVERYTHING, "FpStructInRegistersInfo: struct %s (%u bytes) has no floating fields\n",
+            (!th.IsTypeDesc() ? th.AsMethodTable() : th.AsNativeValueType())->GetDebugClassName(), th.GetSize()));
+        return FpStructInRegistersInfo{};
+    }
```

Comment:
As I alluded to in some other PRs we have the notion of significant padding where it can essentially be considered that all the padding of a struct is covered by fields. If you look at the `getTypeLayout` implementation you can see how it is computed. The JIT takes care to preserve values in bytes not covered by fields in those cases, and the ABI classification will need to do the same.

For example, a structure declaration like
```csharp
private unsafe struct S
{
    public fixed byte Foo[15];
    public float Bar;
}
```

results in underlying metadata that looks like
```csharp
private unsafe struct S
{
  public FooStruct Foo;
  public float Bar;
}

[StructLayout(LayoutKind.Sequential, Size = 15)]
private struct FooStruct
{
  public byte FixedElementField;
}
```

I'm curious how this code ends up classifying a struct like this one for passing.

It seems like RISC-V/LA64 are going to end up with some potentially surprising user behavior here because of these ABI differences when padding is involved, and because it is somewhat ambiguous to the VM/JIT what is (ignorable) padding (or at least not totally obvious to the user).

Reply:
Here's a log from a similar case in JIT/Directed/StructABI/StructABI/StructABI.cs:
```
TID 3f2277: FpStructInRegistersInfo: flattening InlineArray1 (managed, 1 fields)
TID 3f2277: FpStructInRegistersInfo:     flattening <Array>e__FixedBuffer (managed, 1 fields)
TID 3f2277: FpStructInRegistersInfo:      * found field FixedElementField [0..1), type: Byte
TID 3f2277: FpStructInRegistersInfo:          * array has too many elements: 16
```
So it stops the field flattening because there's too many fields (max 2 fields to get passed according to FP calling convention) and returns an empty FpStructInRegistersInfo which means pass according to integer calling convention where there is no notion of "fields", structs are a lump of bits laid out in registers as in memory. But it this case `struct S` is bigger than 16 bytes so it's passed by implicit ref.

Reply:
I see now the handling for `HasImpliedRepeatedField`; that might have to be generalized somewhat. What about the following example?
```
[StructLayout(LayoutKind.Explicit, Size = 20)]
struct S
{
  [FieldOffset(0)]
  public byte FirstByteOfArray;
  [FieldOffset(16)]
  public float FloatField;
}
```

Maybe SysV classification needs some generalization too (I can see it uses `HasImpliedRepeatedFields` too). Or perhaps it is the significant padding computation in `getTypeLayout` that is overly conservative.

Reply:
I think as expected:
```
TID 3f849f: FpStructInRegistersInfo: flattening S (managed, 2 fields)
TID 3f849f: FpStructInRegistersInfo:  * found field FirstByteOfArray [0..1), type: Byte
TID 3f849f: FpStructInRegistersInfo:  * found field FloatField [12..16), type: Single
TID 3f849f: FpStructInRegistersInfo: struct S (16 bytes) can be passed with floating-point calling convention, flags=0x88; IntFloat, sizes={1, 4}, offsets={0, 12}, IntFieldKindMask=Integer
```
Note: I downsized to 16 bytes as I'm on #101796 branch where the condition has not been relaxed for RISC-V. But I think it answers your question (padding via explicit layout).

There are problems with handling `fixed` buffers as the comment in HasImpliedRepeatedField implies. But as far as the classification for RISC-V goes I think it's ok.

Reply:
`getTypeLayout` considers the above to be equivalent to a struct `struct S {public fixed byte Array[16]; public float FloatField; }`. So in that view I don't think the RISC-V classification is ok; IIUC it will silently drop parts of the struct that may contain user data when it gets passed as an argument.

The ABI classification here should match what `getTypeLayout` decides, one way or the other. I'm not sure if it is possible for us to change what it considers significant padding, since that has been in the JIT for a long time now.

Reply:
#71711 has some of the history behind this.

Reply:
BTW I just realized the corner-case where fields are reversed with explicit offsets is not covered, currently it answers "fields overlap, treat as union" and falls back to integer calling convention which is obviously not correct. Ah well, another thing on the todo list..

Reply:
> `getTypeLayout` considers the above to be equivalent to a struct `struct S {public fixed byte Array[16]; public float FloatField; }`. So in that view I don't think the RISC-V classification is ok; IIUC it will silently drop parts of the struct that may contain user data when it gets passed as an argument.

How would the user write data to e.g. `FirstByteOfArray[5]` with standard language features?

Native compilers in general don't consider padding to be preserved while passing, definitely not the RISC-V ABI. That's the whole point of passing a struct according to FP calling convention, as two fields each in one register.

BTW what would be the .NET equivalent for this C/C++ struct?
```C++
struct {
    char i;
    alignas(16) float f;
};
```

> The ABI classification here should match what `getTypeLayout` decides, one way or the other. I'm not sure if it is possible for us to change what it considers significant padding, since that has been in the JIT for a long time now.

That's news to me:/ I don't see `getTypeLayout` used for parameter classification neither on System V nor on RISC-V/LoongArch.

Are there any rules defined for what padding is considered `significant` or just whatever is in the code for `getTypeLayout`? Becuase treating padding as an array does influence the classification for passing so the user would need to be aware of it. The remarks in [StructLayout documentation](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.structlayoutattribute?view=net-8.0) say it's for controlling the layout to pass a type to unmanaged code, i.e. match the layout of the unmanaged type.

Reply:
> How would the user write data to e.g. `FirstByteOfArray[5]` with standard language features?

Using unsafe code. I am not sure of the historical details of how things evolved this way, but I would guess C++/CLI is a large part of it.

> Native compilers in general don't consider padding to be preserved while passing, definitely not the RISC-V ABI. That's the whole point of passing a struct according to FP calling convention, as two fields each in one register.

The difference in .NET is simply what we consider to be the discardable padding. We also have discardable padding, like the last 4 bytes of `Span<T>`. But `ExplicitLayout`/explicit size automatically promote the padding of the struct to be considered as "must be preserved".
I believe most of your example test cases have discardable padding since they do not have `ExplicitLayout`/explicit size, so there I think the ABI classification done here makes sense. But for the example given above things are different, where the ABI classification should essentially be done as if the padding was replaced by explicit `char` arrays of the right size.

> BTW what would be the .NET equivalent for this C/C++ struct?
> 
> ```c++
> struct {
>     char i;
>     alignas(16) float f;
> };
> ```

I am not sure we have an equivalent. I think there are both .NET structs (like the ones with `ExplicitLayout`/`Size`) that are not representable in C/C++; and C/C++ structs (like your example) that are not representable in .NET metadata. @AaronRobinsonMSFT, @jkoritzinsky or @jkotas should know more about the interop story here...

> That's news to me:/ I don't see `getTypeLayout` used for parameter classification neither on System V nor on RISC-V/LoongArch.

I mention `getTypeLayout` because it has the current source of truth of when we consider padding in structs to be significant/required to be preserved. I do not know if the rules in there could be relaxed; just that changing the rules would be breaking, and that users are potentially relying on the values in padding of such structures to be preserved. So in that sense it becomes a problem if the ABI does not match with these rules.

I believe all of our existing ABIs come with rules that will preserve this padding, and thus we have not needed to pay any special attention to it before. I could be wrong about this on SysV, in which case I would consider it a bug.


Ultimately I think `getTypeLayout` and the ABI classification need to agree on what padding in a struct is significant and needs to be preserved, but I do not know whether it would be possible to relax the rules in `getTypeLayout` or not.

Reply:
> Are there any rules defined for what padding is considered `significant` or just whatever is in the code for `getTypeLayout`? Becuase treating padding as an array does influence the classification for passing so the user would need to be aware of it. The remarks in [StructLayout documentation](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.structlayoutattribute?view=net-8.0) say it's for controlling the layout to pass a type to unmanaged code, i.e. match the layout of the unmanaged type, and it doesn't mention any of it.

I don't think we explicitly document this part of the rules anywhere (and as you can see in #71711, the semantics were not clear even to ourselves for a long time). I wouldn't be surprised if `CUSTOMLAYOUT` came to exist quite organically when some issue was noticed where the JIT discarded padding that was expected to be preserved.

Reply:
> I think there are both .NET structs (like the ones with ExplicitLayout/Size) that are not representable in C/C++; and C/C++ structs (like your example) that are not representable in .NET metadata.

Yes, that sounds about right.

Reply:
> I think there are both .NET structs (like the ones with ExplicitLayout/Size) that are not representable in C/C++; and C/C++ structs (like your example) that are not representable in .NET metadata.

Agree.